distributed/worker.py

```diff
@@ -1634,7 +1633,9 @@ def _close(executor, wait):
             # weird deadlocks particularly if the task that is executing in
             # the thread is waiting for a server reply, e.g. when using
             # worker clients, semaphores, etc.
-            if is_python_shutting_down():
+
+            # Are we shutting down the process?
+            if self._is_finalizing() or not threading.main_thread().is_alive():
```

Comment:
This isn't great, but it's the safest way of checking whether the interpreter is shutting down. Threads are being shutdown _before_ `atexit` hooks so we need a mechanism like `is_python_shutting_down` that becomes the first `atexit` hook in `distributed` and lets us know whether we have entered shutdown, or we need to propagate that knowledge during shutdown, or we need to check whether it's safe to use threads where necessary (as done here).

Reply:
what a mess. I'm fine with this